Assets/Mirror/Editor/Weaver/Writers.cs

```diff
@@ -280,5 +285,96 @@ static MethodDefinition GenerateArrayWriteFunc(TypeReference variable, int recur
             worker.Append(worker.Create(OpCodes.Ret));
             return writerFunc;
         }
+
+        static MethodDefinition GenerateArraySegmentWriteFunc(TypeReference variable, int recursionCount)
+        {
+            GenericInstanceType genericInstance = (GenericInstanceType)variable;
+            TypeReference elementType = genericInstance.GenericArguments[0];
+            MethodReference elementWriteFunc = GetWriteFunc(elementType, recursionCount + 1);
+
+            if (elementWriteFunc == null)
+            {
+                return null;
+            }
+
+            string functionName = "_WriteArraySegment_" + elementType.Name + "_";
+            if (variable.DeclaringType != null)
+            {
+                functionName += variable.DeclaringType.Name;
+            }
+            else
+            {
+                functionName += "None";
+            }
+
+            // create new writer for this type
+            MethodDefinition writerFunc = new MethodDefinition(functionName,
+                    MethodAttributes.Public |
+                    MethodAttributes.Static |
+                    MethodAttributes.HideBySig,
+                    Weaver.voidType);
+
+            writerFunc.Parameters.Add(new ParameterDefinition("writer", ParameterAttributes.None, Weaver.CurrentAssembly.MainModule.ImportReference(Weaver.NetworkWriterType)));
+            writerFunc.Parameters.Add(new ParameterDefinition("value", ParameterAttributes.None, variable));
+
+            writerFunc.Body.Variables.Add(new VariableDefinition(Weaver.int32Type));
+            writerFunc.Body.Variables.Add(new VariableDefinition(Weaver.int32Type));
+            writerFunc.Body.InitLocals = true;
+
+            ILProcessor worker = writerFunc.Body.GetILProcessor();
+
+            MethodReference countref = Weaver.ArraySegmentCountReference.MakeHostInstanceGeneric(genericInstance);
+
+            // int length = value.Count;
+            worker.Append(worker.Create(OpCodes.Ldarga_S, (byte)1));
+            worker.Append(worker.Create(OpCodes.Call, countref));
+            worker.Append(worker.Create(OpCodes.Stloc_0));
+
+            
+            // writer.WritePackedInt32(length);
+            worker.Append(worker.Create(OpCodes.Ldarg_0));
+            worker.Append(worker.Create(OpCodes.Ldloc_0));
+            worker.Append(worker.Create(OpCodes.Call, Weaver.NetworkWriterWritePackedInt32));
+
+            // for (int i=0; i< length; i++) {
+            worker.Append(worker.Create(OpCodes.Ldc_I4_0));
+            worker.Append(worker.Create(OpCodes.Stloc_1));
+            Instruction labelHead = worker.Create(OpCodes.Nop);
+            worker.Append(worker.Create(OpCodes.Br, labelHead));
+
+            // loop body
```

Comment:
this loops the parameters? if yes, maybe a comment about it would be more obvious.
I am not 100% sure what this does based on looking at it. or does it loop the arraysegment bytes?

Reply:
This loops the ArraySegment<T>

We already support ArraySegment<byte> which is highly optimized because it reads directly from the NeworkReader buffer instead of making a copy.

This PR is for ArraySegment<T> where T is anything but byte.  

I will add comment

Reply:
I added some comments 10 lines before,  where the loop starts